src/supernova/mod.rs

```diff
@@ -0,0 +1,1202 @@
+    // fold the secondary circuit's instance
+    let (nifs_secondary, (r_U_secondary_folded, r_W_secondary_folded)) = NIFS::prove(
+      ck_secondary,
+      &pp.ro_consts_secondary,
+      &scalar_as_base::<G1>(self.pp_digest),
+      &pp.r1cs_shape_secondary,
+      self.r_U_secondary[0].as_ref().unwrap(),
+      self.r_W_secondary[0].as_ref().unwrap(),
+      &self.l_u_secondary,
+      &self.l_w_secondary,
+    )?;
+
+    // clone and updated running instance on respective circuit_index
+    let r_U_secondary_next = r_U_secondary_folded;
+    let r_W_secondary_next = r_W_secondary_folded;
+
+    let mut cs_primary: SatisfyingAssignment<G1> = SatisfyingAssignment::new();
+    let T = Commitment::<G2>::decompress(&nifs_secondary.comm_T)?;
+    let inputs_primary: SuperNovaAugmentedCircuitInputs<'_, G2> =
+      SuperNovaAugmentedCircuitInputs::new(
+        scalar_as_base::<G1>(self.pp_digest),
+        G1::Scalar::from(self.i as u64),
+        z0_primary,
+        Some(&self.zi_primary),
+        Some(&self.r_U_secondary),
+        Some(&self.l_u_secondary),
+        Some(&T),
+        self.program_counter,
+        G1::Scalar::ZERO,
+      );
+
+    let circuit_primary: SuperNovaAugmentedCircuit<'_, G2, C1> = SuperNovaAugmentedCircuit::new(
+      &pp.augmented_circuit_params_primary,
+      Some(inputs_primary),
+      c_primary,
+      pp.ro_consts_circuit_primary.clone(),
+      self.num_augmented_circuits,
+    );
+
+    let (zi_primary_pc_next, zi_primary) = circuit_primary
+      .synthesize(&mut cs_primary)
+      .map_err(|_| NovaError::SynthesisError)?;
+
+    let (l_u_primary, l_w_primary) = cs_primary
+      .r1cs_instance_and_witness(&pp.r1cs_shape_primary, ck_primary)
+      .map_err(|_| NovaError::UnSat)?;
+
+    // Split into `if let`/`else` statement
+    // to avoid `returns a value referencing data owned by closure` error on `&RelaxedR1CSInstance::default` and `RelaxedR1CSWitness::default`
+    let (nifs_primary, (r_U_primary_folded, r_W_primary_folded)) =
+      if let Some((r_U_primary, r_W_primary)) = self
+        .r_U_primary
+        .get(claim.get_augmented_circuit_index())
+        .unwrap_or(&None)
+        .as_ref()
+        .zip(
+          self
+            .r_W_primary
+            .get(claim.get_augmented_circuit_index())
+            .unwrap_or(&None)
+            .as_ref(),
+        )
+      {
+        NIFS::prove(
+          ck_primary,
+          &pp.ro_consts_primary,
+          &self.pp_digest,
+          &pp.r1cs_shape_primary,
+          r_U_primary,
+          r_W_primary,
+          &l_u_primary,
+          &l_w_primary,
+        )?
+      } else {
+        NIFS::prove(
+          ck_primary,
+          &pp.ro_consts_primary,
+          &self.pp_digest,
+          &pp.r1cs_shape_primary,
+          &RelaxedR1CSInstance::default(ck_primary, &pp.r1cs_shape_primary),
+          &RelaxedR1CSWitness::default(&pp.r1cs_shape_primary),
+          &l_u_primary,
+          &l_w_primary,
+        )?
+      };
```

Comment:
Nit: This might look slightly better as
```rust
match (self.r_U_primary.get(claim.get_augmented_circuit_index()), self.r_W_primary.get(claim.get_augmented_circuit_index())) {
  Some((r_U_primary, r_W_primary)) => ...,
  _ => ...
}
```

Just a thought (what you already have is fine)

Reply:
Adopted the way you suggest which makes code shorter and clean 👍